redaxo/src/core/lib/console/setup/run.php

```diff
@@ -0,0 +1,344 @@
+
+        // ---------------------------------- Step 6 . Create User
+        $io->title('Step 6 of 6 / User');
+
+        $login = null;
+        $password = null;
+
+        if (null === $input->getOption('admin-username') || null === $input->getOption('admin-password')) {
+            $user = rex_sql::factory();
+            $user
+                ->setTable(rex::getTable('user'))
+                ->select();
+
+            $skipUserCreation = false;
+            if ($user->getRows()) {
+                $skipUserCreation = $io->confirm('Users already exists. Skip user creation?');
+            }
+            $passwordPolicy = rex_backend_password_policy::factory(rex::getProperty('password_policy', []));
+
+            if (!$skipUserCreation) {
+                $io->section('Create administrator account');
+                $login = $io->ask('Username', null, static function ($login) {
+                    $user = rex_sql::factory();
+                    $user
+                        ->setTable(rex::getTable('user'))
+                        ->setWhere(['login' => $login])
+                        ->select();
+
+                    if ($user->getRows()) {
+                        throw new InvalidArgumentException(sprintf('User "%s" already exists.', $login));
+                    }
+                    return $login;
+                });
+                $password = $io->askHidden('Password', static function ($password) use ($passwordPolicy) {
+                    if (true !== $msg = $passwordPolicy->check($password)) {
+                        throw new InvalidArgumentException($msg);
+                    }
+
+                    return $password;
+                });
+            }
+        } else {
+            $login = $input->getOption('admin-username');
+            $password = $input->getOption('admin-password');
+        }
+
+        if ($login && $password) {
+            $user = rex_sql::factory();
+            $user
+                ->setTable(rex::getTable('user'))
+                ->setWhere(['login' => $login])
+                ->select();
+
+            if ($user->getRows()) {
+                throw new InvalidArgumentException(sprintf('User "%s" already exists.', $login));
+            }
+
+            $user = rex_sql::factory();
+            $user->setTable(rex::getTablePrefix() . 'user');
+            $user->setValue('login', $login);
+            $user->setValue('password', rex_backend_login::passwordHash($password));
+            $user->setValue('admin', 1);
+            $user->addGlobalCreateFields('console');
+            $user->addGlobalUpdateFields('console');
+            $user->setValue('status', '1');
+            $user->insert();
+
+            $io->success(sprintf('User "%s" successfully created.', $login));
+        }
+
+        // ---------------------------------- last step. save config
+
+        $config['setup'] = false;
+        if (!rex_file::putConfig($configFile, $config)) {
```

Comment:
Bin mir unsicher, ob es gut ist, erst hier ganz am Ende zu speichern. Denn die anderen Steps führen ihre Parts ja teils auch direkt aus (DB init etc.), daher eventuell besser auch die Config nach jedem Schritt speichern, wo sich was ändert?

Reply:
klar kann man zwischendurch speichern, die Frage ist, was das für ein Vorteil hat. Das Setup würde ja trotzdem immer wieder von vorne beginnen.
Einzige Stelle wäre IMO zum Zeitpunkt des DB inits.

Reply:
> die Frage ist, was das für ein Vorteil hat

Ja, das ist eine gute Frage. So richtig weiß ich es auch nicht. :D
Also man hätte zumindest den Vorteil, dass die eingegebenen Werte bei erneuter Durchführung als Defaults vorgeschlagen werden, und man nur noch Enter drücken braucht, oder?

Und was mich ansonsten so ein bisschen störte: Angenommen beim User-Erstellen läuft was schief. Dann wurde die DB schon initialisiert etc, aber die Verbindung dazu ist noch gar nicht hinterlegt in REDAXO.

Aber ich sehe auch Vorteile bei deiner Variante. Also bin mir unsicher.

@staabm Hast du eine Meinung dazu?

Reply:
Ausm bauch: macht es sinn eine Transaktion zu verwenden um sodass das command immer alles oder nix ist?

Reply:
transaction wäre cool aber es gibt statements (wie z.B. `Create table`), die einen implicit commit machen.

siehe https://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html

Reply:
ahh stimmt. transactionales DDL geht erst mit mysql 8.x

Reply:
ok, da somit keine transaction, was sagt dir dann der bauch @staabm? :)

Reply:
Ich würde erstmal nur abschließend speichern.. ich vermute dass die meisten user aus automatismen damit arbeiten und es nur wenige user gibt

Reply:
jupp ok